lib/mongoid/traversable.rb

```diff
@@ -27,7 +27,7 @@ def _parent=(p)
     #
     # @return [ Array<Document> ] All child documents in the hierarchy.
     def _children
-      @__children ||= collect_children
```

Comment:
Yes, thats indeed a problem, however I am not sure removing the memoization is the right solution.
We can clear that once we do a new change on the root document.